DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp

```diff
@@ -0,0 +1,552 @@
+      } else if (arg.find("--iterations=") == 0) {
+        iterations = std::max(2, atoi(str_after_equals.c_str()) + 1);
+      } else {
+        std::cout << "WARNING: ignoring unknown argument '" << arg << "'\n";
+      }
+    }
+  }
+
+  // check the chunks
+  if (chunks <= 0) {
+    std::cerr << "ERROR: 'chunks' must be greater than 0\n";
+    return 1;
+  }
+
+  // check the chunk size
+  if (chunk_count <= 0) {
+    std::cerr << "ERROR: 'chunk_count' must be greater than 0\n";
+    return 1;
+  }
+
+  // check inflight_kernels
+  if (inflight_kernels <= 0) {
+    std::cerr << "ERROR: 'inflight_kernels' must be positive\n";
+    return 1;
+  }
+
+  // check the number of iterations
+  if (iterations <= 0) {
+    std::cerr << "ERROR: 'iterations' must be positive\n";
+    return 1;
+  }
+
+  // compute the total number of elements
+  size_t total_count = chunks * chunk_count;
+
+  std::cout << "# Chunks:             " << chunks << "\n";
+  std::cout << "Chunk count:          " << chunk_count << "\n";
+  std::cout << "Total count:          " << total_count << "\n";
+  std::cout << "Iterations:           " << iterations-1 << "\n";
+  std::cout << "\n";
+
+  bool passed = true;
+
+  try {
+    // device selector
+#if defined(FPGA_EMULATOR)
+    INTEL::fpga_emulator_selector selector;
+#else
+    INTEL::fpga_selector selector;
+#endif
+
+    // queue properties to enable profiling
+    auto prop_list = property_list{ property::queue::enable_profiling() };
+
+    // create the device queue
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // make sure the device supports USM host allocations
+    device d = q.get_device();
+    if (!d.get_info<info::device::usm_host_allocations>()) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations\n";
+      return 1;
+    }
+
+    // the USM input and output data
+    Type *in, *out;
+    if ((in = malloc_host<Type>(total_count, q)) == NULL) {
+      std::cerr << "ERROR: could not allocate space for 'in'\n";
+      std::terminate();
+    }
+    if ((out = malloc_host<Type>(total_count, q)) == NULL) {
+      std::cerr << "ERROR: could not allocate space for 'out'\n";
+      std::terminate();
+    }
+
+    // generate the random input data
+    // NOTE: by generating all of the data ahead of time, we are essentially
+    // assuming that the producer of data (producing data for the FPGA to
+    // consume) has infinite bandwidth. However, if the producer of data cannot
+    // produce data faster than our FPGA can consume it, the CPU producer will
+    // bottleneck the total throughput of the design.
+    std::generate_n(in, total_count, [] { return Type(rand() % 100); });
+
+    // a lambda function to validate the results
+    auto validate_results = [&]() {
+      for (size_t i = 0; i < total_count; i++) {
+        auto comp = (in[i] == out[i]);
+        for (auto j = 0; j < comp.get_count(); j++) {
```

Comment:
I am lost here... What is the type of `comp`?

Reply:
I found it... `sycl::long8`.

Reply:
I thought you liked `auto` ;)?

Reply:
Indeed, I am a fan :-) but I was puzzled by not having `comp` being directly a `bool`...